### PR TITLE
Add support for requirements files in `uv run`

### PR DESCRIPTION
**Pull request by zanieb**
Closes https://github.com/astral-sh/uv/issues/4824.

### CODE REVIEW

**Comment by zanieb** [comment]
Terrifyingly easy to add. I want to add some more test coverage and I'm sure there are some niche behaviors I need to explore.

File: crates/uv-cli/src/lib.rs [hunk 1]
```diff
@@ -1781,6 +1781,37 @@ pub struct RunArgs {
...
     pub python: Option<String>,
+
+    /// Require all packages listed in the given `requirements.txt` files.
+    ///
+    /// If a `pyproject.toml`, `setup.py`, or `setup.cfg` file is provided, `uv` will
+    /// extract the requirements for the relevant project.
+    ///
+    /// If `-` is provided, then requirements will be read from stdin.
+    #[arg(long, short, group = "sources", value_parser = parse_file_path)]
+    pub requirement: Vec<PathBuf>,
```
Comment (zanieb) [inline crates/uv-cli/src/lib.rs:1792]: Should we say `-r` / `--with-requirements` instead?

File: crates/uv/src/commands/project/run.rs [hunk 2]
```diff
@@ -56,6 +58,27 @@ pub(crate) async fn run(
         warn_user_once!("`uv run` is experimental and may change without warning.");
     }
 
+    // These cases seem quite complex because it changes the current package — let's just
+    // ban it entirely for now
```
Comment (zanieb) [inline crates/uv/src/commands/project/run.rs:62]: Alternatively, `uv run -r` could imply `--isolated` (which might make sense but is more limiting?)

Comment (charliermarsh) [reply crates/uv/src/commands/project/run.rs:62]: Am I correct that the _code_ would work, and here, you just mean that semantically `-r foo/bar/pyproject.toml` _should_ use `foo/bar` as the project?

**Review by charliermarsh — APPROVED** [review]
Seems fine (it's good that it was easy!). Let's also add to `uv tool run`? What about `uv tool install`...?

**Comment by charliermarsh** [comment]
Also closes https://github.com/astral-sh/uv/issues/4824.

File: crates/uv/src/commands/project/run.rs (hunk 2 above)
Comment (zanieb) [reply crates/uv/src/commands/project/run.rs:62]: Yeah the code should work as-is, it's just weird since it may _imply_ a change of project.

Comment (zanieb) [reply crates/uv/src/commands/project/run.rs:62]: Like.. you should use `--with ./foo/bar` instead, right?

**Comment by charliermarsh** [comment]
We don't accept `--override foo==1.0.0` in the `pip` API -- is it necessary to support here? It seems fine to me to be consistent with the `pip` API on those.

**Comment by charliermarsh** [comment]
Or, at least, we definitely don't need to support that to start IMO. And if we want to support it, we should probably support it in the pip API too.

**Comment by zanieb** [comment]
Yeah I don't think that's in scope for this pull request, just want to nail down the plan so we can make sure it's coherent.

**Comment by charliermarsh** [comment]
I think I'd be ok with `--requirements`, `--overrides`, and `--constraints` (with the short options too) -- it just feels the most intuitive coming from the other APIs. `--with-*` is also ok, but we'd probably end up wanting aliases for the shorter versions?

**Comment by charliermarsh** [comment]
@zanieb - I'm happy to finish this PR if we can get consensus on the names, no trouble.

**Comment by zanieb** [comment]
@charliermarsh I am comfortable finishing it up this week

File: crates/uv/src/commands/project/run.rs [hunk 3]
```diff
@@ -281,6 +310,7 @@ pub(crate) async fn run(
     // any `--with` requirements, and we already have a base environment, then there's no need to
     // create an additional environment.
     let skip_ephemeral = base_interpreter.as_ref().is_some_and(|base_interpreter| {
+        // No additional requrements
```
Comment (charliermarsh) [inline crates/uv/src/commands/project/run.rs:313]: Typo :D

File: crates/uv/src/commands/project/run.rs [hunk 4]
```diff
@@ -363,35 +400,28 @@ pub(crate) async fn run(
...
-                    &state,
-                    preview,
-                    connectivity,
-                    concurrency,
-                    native_tls,
-                    cache,
-                    printer,
+        match spec {
+            None => Some(venv),
+            Some(spec) if spec.is_empty() => Some(venv),
```
Comment (charliermarsh) [inline crates/uv/src/commands/project/run.rs:405]: You might be able to simplify this with a `spec.filter(|spec| !spec.is_empty())` somewhere.

**Review by charliermarsh — APPROVED** [review]

**Comment by charliermarsh** [comment]
As discussed I'll take this over and try to get it merged today.

**Comment by zanieb** [comment]
Note that we're using plural `--constraints`, `--requirements` and `--overrides` here to make room for the singular interface in the future. However, this differs from the `uv pip` interface which uses singular names to refer to files.

**Comment by charliermarsh** [comment]
We decided to narrow the scope to requirements files for now. Constraints and overrides require that we resolve _alongside_ the base environment, diff the result, and install any changes into the ephemeral environment. Which is doable, but a lot more complex.

File: crates/uv-cli/src/lib.rs [hunk 5]
```diff
@@ -1833,6 +1833,14 @@ pub struct RunArgs {
     #[arg(long)]
     pub with: Vec<String>,
 
+    /// Run with all packages listed in the given `requirements.txt` files.
+    ///
+    /// Using `pyproject.toml`, `setup.py`, or `setup.cfg` files is not allowed.
+    ///
+    /// If `-` is provided, then requirements will be read from stdin.
```
Comment (zanieb) [inline crates/uv-cli/src/lib.rs:1840]: This seems dubious here, is it easy to disable?

Comment (charliermarsh) [reply crates/uv-cli/src/lib.rs:1840]: I don't think so...

File: crates/uv/src/commands/project/run.rs [hunk 6]
```diff
@@ -292,6 +310,14 @@ pub(crate) async fn run(
...
         }
 
+        // `SitePackages::satisfies` doesn't handle overrides yet, and constraints are only enforced
+        // on `--with` requirements instead of the project requirements, so we perform a full
+        // resolution to ensure things are up-to-date.
+        // TODO(charlie): Support constraints and overrides in `uv run`.
+        if !(spec.constraints.is_empty() && spec.overrides.is_empty()) {
+            return false;
+        }
+
```
Comment (zanieb) [inline crates/uv/src/commands/project/run.rs:320]: This case can't occur now, right? Still worth including?

File: crates/uv-cli/src/lib.rs (hunk 5 above)
Comment (charliermarsh) [reply crates/uv-cli/src/lib.rs:1840]: Checking

File: crates/uv/src/commands/project/run.rs (hunk 6 above)
Comment (charliermarsh) [reply crates/uv/src/commands/project/run.rs:320]: I figured it was worth it with a TODO. I can clarify more in the comments.

File: crates/uv-cli/src/lib.rs (hunk 5 above)
Comment (charliermarsh) [reply crates/uv-cli/src/lib.rs]: Fixed

File: crates/uv/src/commands/project/run.rs (hunk 6 above)
Comment (charliermarsh) [reply crates/uv/src/commands/project/run.rs]: Removed

**Review by zanieb — COMMENTED** [review]
"Approved"

**Comment by zanieb** [comment]
Thanks for taking this over!

**Pull request merged**